flink-end-to-end-tests/src/main/java/org/apache/flink/streaming/tests/StickyAllocationAndLocalRecoveryTestJob.java

```diff
@@ -0,0 +1,451 @@
+					for (MapperSchedulingAndFailureInfo testInfo : iterable) {
+
+						completeInfo.add(testInfo);
+
+						if (taskNameWithSubtasks.equals(testInfo.taskNameWithSubtask)) {
+							infoForThisTask = testInfo;
+						}
+
+						if (testInfo.killedJvm) {
+							killedJvmPids.add(testInfo.jvmPid);
+						}
+					}
+				}
+
+				Preconditions.checkNotNull(infoForThisTask, "Expected to find info here.");
+
+				if (!isScheduledToCorrectAllocation(infoForThisTask, allocationID, killedJvmPids)) {
+					allocationFailureMessage = String.format(
+						"Sticky allocation test failed: Subtask %s in attempt %d was rescheduled from allocation %s " +
+							"on JVM with PID %d to unexpected allocation %s on JVM with PID %d.\n" +
+							"Complete information from before the crash: %s.",
+						runtimeContext.getTaskNameWithSubtasks(),
+						runtimeContext.getAttemptNumber(),
+						infoForThisTask.allocationId,
+						infoForThisTask.jvmPid,
+						allocationID,
+						thisJvmPid,
+						completeInfo);
+				}
+			}
+
+			// We determine which of the subtasks will produce the artificial failure
+			boolean failingTask = shouldTaskFailForThisAttempt();
+
+			// We take note of all the meta info that we require to check sticky scheduling in the next re-attempt
+			this.currentSchedulingAndFailureInfo = new MapperSchedulingAndFailureInfo(
+				failingTask,
+				failingTask && killTaskOnFailure,
+				thisJvmPid,
+				runtimeContext.getTaskNameWithSubtasks(),
+				allocationID);
+
+			schedulingAndFailureState.clear();
+			schedulingAndFailureState.add(currentSchedulingAndFailureInfo);
+		}
+
+		@Override
+		public void notifyCheckpointComplete(long checkpointId) {
+			// we can only fail the task after at least one checkpoint is completed to record progress.
+			failTask = currentSchedulingAndFailureInfo.failingTask;
+		}
+
+		private boolean shouldTaskFailForThisAttempt() {
+			RuntimeContext runtimeContext = getRuntimeContext();
+			int numSubtasks = runtimeContext.getNumberOfParallelSubtasks();
+			int subtaskIdx = runtimeContext.getIndexOfThisSubtask();
+			int attempt = runtimeContext.getAttemptNumber();
+			return (attempt % numSubtasks) == subtaskIdx;
+		}
+
+		private boolean isScheduledToCorrectAllocation(
+			MapperSchedulingAndFailureInfo infoForThisTask,
+			String allocationID,
+			Set<Integer> killedJvmPids) {
+
+			return (infoForThisTask.allocationId.equals(allocationID)
+				|| killedJvmPids.contains(infoForThisTask.jvmPid));
+		}
+	}
+
+	private static int getJvmPid() throws Exception {
+		java.lang.management.RuntimeMXBean runtime =
+			java.lang.management.ManagementFactory.getRuntimeMXBean();
+		java.lang.reflect.Field jvm = runtime.getClass().getDeclaredField("jvm");
+		jvm.setAccessible(true);
+		sun.management.VMManagement mgmt =
+			(sun.management.VMManagement) jvm.get(runtime);
+		java.lang.reflect.Method pidMethod =
+			mgmt.getClass().getDeclaredMethod("getProcessId");
+		pidMethod.setAccessible(true);
```

Comment:
Please attribute this code snippet if it comes from https://stackoverflow.com/a/12066696/4815083

Reply:
AFAIK, the SO license is not compatible with the ASL. So if it comes from SO, we cannot use it.

Reply:
Then we have a problem with other code parts as well, if I'm not mistaken.

Reply:
Stack Overflow is [licensed CC BY-SA 3.0](https://stackoverflow.com/help/licensing) which is [Apache approved](https://www.apache.org/legal/resolved.html#cc-sa). Our attributions do need to be more than a link.

Reply:
Thanks for clarifying the issue @greghogan.

Reply:
👍 added attributions. In case of license problems, I wonder how this would work for such trivial code that is just calling three getters in an API?